app/main.py

```diff
@@ -1,34 +1,143 @@
+        self.get_axis()
+
+        self.decks: [Deck] = []  # Deck/Decks of every Ship
+
+        """."""
+        if self.direction == "single_point":
+            self.decks.append(Deck(row=start[0], column=start[1]))
+            # self.len = 1
+            # print(self.len)
+
+        """---->"""
+        if self.direction == "x":
+            self.decks = [Deck(row=self.start[0], column=coord) for coord
+                          in range(self.start[1], self.end[1] + 1)]
+            # self.len = len(self.decks)
+            # print(self.len)
+
+        """↓"""
+        if self.direction == "y":
+            self.decks = [Deck(row=coord, column=self.start[1]) for coord
+                          in range(self.start[0], self.end[0] + 1)]
+            # self.len = len(self.decks)
+            # print(self.len)
+
+    def get_axis(self):  # custom
+        if self.start == self.end:
+            self.direction = "single_point"  # □
+            return
+        if self.start[0] == self.end[0]:
+            self.direction = "x"  # □ □ □
+            return
+        self.direction = "y"  # □
+        # ______________________□
+        # ______________________□
+
+    def get_deck(self, row: int, column: int) -> Deck:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+        # Find the Deck instance by coords
+
+    def fire(self, row: int, column: int) -> None:
+        deck = self.get_deck(row, column)
+        if deck:
+            deck.is_alive = False
+            self.decks.remove(deck)
+            if len(self.decks) == 0:
+                self.is_drowned = True
+
+    def __repr__(self):
+        return tuple(self.start, self.end)
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
+    def __init__(self, ships: list[tuple]):  # Whole game
+        self.field = {}  # Create a dict `self.field`. # ((), (), ... ()) : self.Ship
+        counter = 1
+        for i in ships:  # debug
+            ship = Ship(start=i[0], end=i[1])
+            # print(f"Ship # {counter}:\n{ship}\n")  # debug
+            counter += 1  # debug
+            for deck in ship.decks:
+                self.field[deck] = ship
+        # print(self.field)
+        # self.validate_input()
+        # self.print_field()
+        # print(self.field)
+        self.fire((7, 7))
 
     def fire(self, location: tuple):
+        for coord, ship in self.field.items():
+            point = (coord.row, coord.column)
+            if point == location:
+                ship.fire(*location)
+                if ship.is_drowned is True:
+                    return "Sunk!"
+                else:
+                    return "Hit!"
+        return "Miss!"
```

Comment:
Loop is not needed. Just check:
if location in self.field:
.....

Reply:
how can i check it without iteration if my key : value pairs looks like:
`<__main__.Deck object at 0x00000231EA4C2B30>: <__main__.Ship object at 0x00000231EA4C2B00>`
I tried:
`deck = Deck(*location)` but
`self.field[deck]` raised `Keyerror`, even with same hash of deck obj 
`hash((self.row, self.column, self.is_alive))`